### PR TITLE
4.x: Ability to Inject MockBeans in Helidon #7694

### PR DESCRIPTION
**Pull request by jbescos**
### Description
https://github.com/helidon-io/helidon/issues/7694

### Documentation

This feature provides a simple mechanism for replacing CDI bean with mock instances created with Mockito.
It is similar to Spring's `@MockBean`, in fact it uses the same name.

The introduced API consist of a new Maven module `io.helidon.microprofile.testing:helidon-microprofile-testing-mocking` that provides one annotation `@MockBean`.

The annotation is designed to be used on fields of test classes annotated with `@HelidonTest`, the implementation relies only on CDI thus it works with either JUnit or TestNG.

The annotation has a parameter "answer" that can be used to set the default answer for the mocked beans.

See a snippet below:
```java
@HelidonTest
@AddBean(TestMockBeanAnswer.Resource.class)
@AddBean(TestMockBeanAnswer.Service.class)
class TestMockBeanAnswer {

    @MockBean(answer = Answers.CALLS_REAL_METHODS)
    private Service service;
    @Inject
    private WebTarget target;

    @Test
    void injectionTest() {
        String response = target.path("/test").request().get(String.class);
        assertThat(response, is("Not Mocked"));
        Mockito.when(service.test()).thenReturn("Mocked");
        response = target.path("/test").request().get(String.class);
        assertThat(response, is("Mocked"));
    }

    @Path("/test")
    public static class Resource {

        @Inject
        private Service service;

        @GET
        public String test() {
            return service.test();
        }
    }

    static class Service {

        String test() {
            return "Not Mocked";
        }

    }
}
```

The mocked beans of a test class can be customized with a CDI producer method.
E.g.

```java
@Produces
MockSettings mockSettings() {
    return Mockito.withSettings().defaultAnswer(Answers.RETURNS_DEFAULTS);
}
```

### CODE REVIEW

**Comment by jbescos** [comment]
Setting it as a draft for now. I have a couple of CDI questions:

Question 1:

Currently `@MockTest` requires also` @Inject`:
```
    @Inject
    @MockBean
    private Service service;
```
Is there any way to not require `@Inject`?.

Question 2:

It does not work if the bean already exists in CDI. For example, this will not work because the service will be defined twice:
```
@HelidonTest
@AddBean(Service.class)
public class TestMockBean {

    @Inject
    @MockBean
    private Service service;
```
I didn't find how to replace the existing bean defined by the mocked one.

**Comment by romain-grecourt** [comment]
> I didn't find how to replace the existing bean defined by the mocked one.
I'm not a CDI expert, but can you set the mocked bean as an alternative ? That's how the current solution (#8339) works.

**Comment by jbescos** [comment]
> > I didn't find how to replace the existing bean defined by the mocked one.
> 
> I'm not a CDI expert, but can you set the mocked bean as an alternative ? That's how the current solution (#8339) works.

I think that requires a producer, and I cannot use a producer because you need to define the type. I will try it further.

**Comment by jbescos** [comment]
For TestNG it seems we cannot use the @Inject in constructor, so the `@MockBean` applies only in fields.

For example, this simple test:
@HelidonTest
```java
class TestMockBeanField {

    private final WebTarget target;

    @Inject
    public TestMockBeanField(WebTarget target) {
        this.target = target;
    }

    @Test
    void test() {
    }
}
```

It fails with:

```
org.testng.TestNGException: 
An error occurred while instantiating class io.helidon.microprofile.tests.testing.testng.TestMockBeanField. Check to make sure it can be instantiated/accessed.
	at org.testng.internal.objects.SimpleObjectDispenser.createInstance(SimpleObjectDispenser.java:113)
	at org.testng.internal.objects.SimpleObjectDispenser.dispense(SimpleObjectDispenser.java:40)
	at org.testng.internal.objects.GuiceBasedObjectDispenser.dispense(GuiceBasedObjectDispenser.java:28)
	at org.testng.internal.ClassImpl.getDefaultInstance(ClassImpl.java:106)
	at org.testng.internal.ClassImpl.getInstances(ClassImpl.java:136)
	at org.testng.TestClass.getInstances(TestClass.java:129)
	at org.testng.TestClass.initTestClassesAndInstances(TestClass.java:109)
	at org.testng.TestClass.init(TestClass.java:101)
	at org.testng.TestClass.<init>(TestClass.java:66)
	at org.testng.TestRunner.initMethods(TestRunner.java:483)
	at org.testng.TestRunner.init(TestRunner.java:356)
	at org.testng.TestRunner.init(TestRunner.java:309)
	at org.testng.TestRunner.<init>(TestRunner.java:184)
	at org.testng.SuiteRunner$DefaultTestRunnerFactory.newTestRunner(SuiteRunner.java:652)
	at org.testng.SuiteRunner.init(SuiteRunner.java:224)
	at org.testng.SuiteRunner.<init>(SuiteRunner.java:116)
	at org.testng.TestNG.createSuiteRunner(TestNG.java:1375)
	at org.testng.TestNG.createSuiteRunners(TestNG.java:1349)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1191)
	at org.testng.TestNG.runSuites(TestNG.java:1114)
	at org.testng.TestNG.run(TestNG.java:1082)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:155)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:102)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:91)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:137)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
```

File: microprofile/tests/testing/junit5/src/test/java/io/helidon/microprofile/tests/testing/junit5/TestMockBeanField.java
Comment (romain-grecourt) [inline microprofile/tests/testing/junit5/src/test/java/io/helidon/microprofile/tests/testing/junit5/TestMockBeanField.java:67]: How about we add an annotation attribute for the default answer ?

File: microprofile/testing/junit5/pom.xml
Comment (romain-grecourt) [inline microprofile/testing/junit5/pom.xml:57]: Can you consult @tomas-langer about adding this dependency.

**Review by romain-grecourt — COMMENTED** [review]
Looks nice so far.
We need to update the docs as well.

File: microprofile/tests/testing/junit5/src/test/java/io/helidon/microprofile/tests/testing/junit5/TestMockBeanField.java
Comment (jbescos) [reply microprofile/tests/testing/junit5/src/test/java/io/helidon/microprofile/tests/testing/junit5/TestMockBeanField.java:67]: I didn't go for it because answer is only one of many Mockito settings. I thought if I am adding the answer, I would need to cover every other setting with annotation attributes, otherwise people will start asking why answer is supported in annotation, but other things are not.

File: microprofile/testing/junit5/pom.xml
Comment (jbescos) [reply microprofile/testing/junit5/pom.xml:57]: It is necessary to compile `helidon-microprofile-testing-junit5`. Users don't need to include mockito dependencies as long as they don't add `@MockBean`. In case they add `@MockBean` without having mockito dependency they will have a runtime error ClassNotFoundException.

I would personally remove every provided scope in the pom.

Comment (romain-grecourt) [reply microprofile/testing/junit5/pom.xml:57]: `HelidonJunitExtension` imports mockito so it's effectively required to have it on the test class-path.
I'm basically asking if we should modularize this. 

@tomas-langer please chime in.

File: microprofile/tests/testing/junit5/src/test/java/io/helidon/microprofile/tests/testing/junit5/TestMockBeanField.java
Comment (romain-grecourt) [reply microprofile/tests/testing/junit5/src/test/java/io/helidon/microprofile/tests/testing/junit5/TestMockBeanField.java:67]: The default answer is one of the (if not THE) main setting, and `Answers` is an enum so it's compatible with an attribute.

The whole point of `@MockBean` is to avoid forcing users into CDI based mechanism, something as basic as the default answer should be supported without having to use `@Produces`.

File: microprofile/testing/junit5/pom.xml
Comment (tomas-langer) [reply microprofile/testing/junit5/pom.xml:57]: Import does not matter, code path that uses the imported type matters. I will check the code.

Comment (tomas-langer) [reply microprofile/testing/junit5/pom.xml:57]: If the dependency needs to be added by the user (i.e. we expect the user already has it on their classpath), it should be `provided`. If it is not required to run this code, it should be `<optional>true</optional>` instead.

File: microprofile/testing/junit5/src/main/java/io/helidon/microprofile/testing/junit5/HelidonJunitExtension.java
Comment (tomas-langer) [inline microprofile/testing/junit5/src/main/java/io/helidon/microprofile/testing/junit5/HelidonJunitExtension.java:623]: This is the code that would fail at runtime, if you use `@MockBean` and you do not have mockito on the classpath. We should never do this.
So either we need extensibility and have a mockito support for `MockBean`, or fallback to some other approach (such as creating a no-op proxy) if it is not available.

File: microprofile/testing/testng/pom.xml
Comment (tomas-langer) [inline microprofile/testing/testng/pom.xml:50]: Again, provided scope is not good here.
We expect TestNG (in this case), and Helidon as provided, as these must be present to even use this module, and users add them to their projects for testing.
Mockito is not an auto-include for testing, so we should not do this.

File: microprofile/testing/testng/src/main/java/io/helidon/microprofile/testing/testng/HelidonTestNgListener.java
Comment (tomas-langer) [inline microprofile/testing/testng/src/main/java/io/helidon/microprofile/testing/testng/HelidonTestNgListener.java:710]: This type already exists:
`jakarta.enterprise.inject.literal.InjectLiteral` there should be no need to create it here or in Junit extension, as we expect CDI to be on classpath (as otherwise this module has nothing to do)

File: microprofile/testing/junit5/pom.xml
Comment (romain-grecourt) [reply microprofile/testing/junit5/pom.xml:57]: I.e. we don't have to modularize, we only need to ensure that the normal code path doesn't fail because of the absence of Mockito.

**Review by tomas-langer — COMMENTED** [review]
This looks like a great feature, but the hard dependency on `mockito` is something we probably do not desire.
It seems to me that the whole feature is implemented as a CDI extension. Can we maybe create a new module (depending on mockito), that would add the CDI extension and the annotation, and that could be used with both JUnit5 and testNG extension, because it is "just" a CDI portable extension?
i.e. `helidon-microprofile-testing-mock-beans`?

File: dependencies/pom.xml
Comment (jbescos) [inline dependencies/pom.xml:509]: @barchetta to review this

File: microprofile/tests/testing/junit5/src/test/java/io/helidon/microprofile/tests/testing/junit5/TestMockBeanField.java
Comment (jbescos) [reply microprofile/tests/testing/junit5/src/test/java/io/helidon/microprofile/tests/testing/junit5/TestMockBeanField.java:67]: Done

File: microprofile/testing/mock-beans/pom.xml
Comment (tomas-langer) [inline microprofile/testing/mock-beans/pom.xml:38]: This should depend on CDI API only (not on Helidon MP, unless you actually use it).
Also this must be a `provided` dependency, as the application must already have it (otherwise there would be no beans to mock)

File: microprofile/testing/mock-beans/src/main/java/module-info.java
Comment (tomas-langer) [inline microprofile/testing/mock-beans/src/main/java/module-info.java:31]: please use fully qualified class name for `with`

**Review by tomas-langer — COMMENTED** [review]
Looks good now, leaving to Romain to review.

**Review by tomas-langer — COMMENTED** [review]
Looks good now, let's wait for library approval.

File: dependencies/pom.xml
Comment (barchetta) [reply dependencies/pom.xml:509]: Do we use this dependency directly (I do not see where we do) or are we managing the version of a transitive dependency? If the later, is this for dependency convergence? Forced upgrade? I don't see an issue here, I just want to understand why. Thx.

File: microprofile/testing/mock-beans/pom.xml
Comment (romain-grecourt) [reply microprofile/testing/mock-beans/pom.xml:45]: We still expect direct mockito usage, so having it as provided can work. However we do still want to manage the version for convenience.

File: microprofile/testing/mock-beans/pom.xml
Comment (romain-grecourt) [inline microprofile/testing/mock-beans/pom.xml:29]: `mock-beans` feels too specific, how about `mocking` instead ?

I.e
- gav: `io.helidon.microprofile.testing:helidon-microprofile-testing-mocking`
- package: `io.helidon.microprofile.testing.mocking`

File: dependencies/pom.xml
Comment (tomas-langer) [reply dependencies/pom.xml:509]: Dependency convergence - mockito and hibernate.

File: microprofile/testing/mock-beans/pom.xml
Comment (tomas-langer) [reply microprofile/testing/mock-beans/pom.xml:45]: Making this a `provided` dependency would not be the ideal solution I think. Customer can still version managed `mockito-core` to any compatible version they need.
My reasoning - Junit and Helidon are expected dependencies, as you are writing tests for a Helidon application.
Mockito is not "auto-include" for tests. 

So we can use `provided`, but it may confuse the users, as the feature would fail to compile unless they explicitly add `mockito` to their project as well.

File: microprofile/testing/mock-beans/pom.xml
Comment (tomas-langer) [reply microprofile/testing/mock-beans/pom.xml:29]: The `mock-beans` seems very specific, agreed. Maybe this deserves one more cycle to change the name.

File: microprofile/testing/mock-beans/pom.xml
Comment (romain-grecourt) [reply microprofile/testing/mock-beans/pom.xml:45]: I'm assuming that when using `@MockBean` users also always use Mockito explicitly.

E.g.
```java
import io.helidon.microprofile.testing.junit5.HelidonTest;
import io.helidon.microprofile.testing.mockbeans.MockBean;

import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

@HelidonTest
class MyTest {

    @MockBean
    private MyService service;

    @Inject
    private WebTarget target;

    @Test
    void testMyService() {
        Mockito.when(service.test()).thenReturn("Mocked");
        response = target.path("/test").request().get(String.class);
        assertThat(response, is("Mocked"));
    }
}
```

One may not actually not use Mockito directly (even though it doesn't make sense) in which case you'd have a runtime error. This is probably a good reason not to use `provided`.

Otherwise, users would have to resolve the `org.mockito.Mockito` import and would naturally add the mockito Maven dependency.

Comment (jbescos) [reply microprofile/testing/mock-beans/pom.xml]: To me it looks equally confusing that Helidon will not bring automatically junit or mockito as a transitive dependency. So it is consistent that if junit is not coming as a transitive dependency, then mockito is also not coming as a transitive dependency.

I am setting it as provided.

File: microprofile/testing/mock-beans/pom.xml
Comment (jbescos) [reply microprofile/testing/mock-beans/pom.xml]: I agree. Done.

**Review by barchetta — DISMISSED** [review]
Dependencies look good to me. Thanks for all the work on this!

**Review by romain-grecourt — COMMENTED** [review]
Looks good, please update the PR description to fill-in the documentation section:
- describe the feature (what it solves)
- describe the API (annotation + new GAV)
- add sample snippet

Let's also file an issue to create the actual docs for this feature.

**Comment by jbescos** [comment]
> Looks good, please update the PR description to fill-in the documentation section:
> 
>     * describe the feature (what it solves)
> 
>     * describe the API (annotation + new GAV)
> 
>     * add sample snippet
> 
> 
> Let's also file an issue to create the actual docs for this feature.

Do we re-use this issue?: https://github.com/helidon-io/helidon/issues/8339

**Comment by romain-grecourt** [comment]
> Do we re-use this issue?: #8339

Let's file a separate issue as #8339 has more to it than strictly document this new feature.
If we want to proceed with #8338 we have to figure out where it fits in the docs.

> I didn't work in any documentation issue before, is there any guide?

No, however  adding one more document is relatively easy as there is a lot of precedence.

---

The bullet items I provided was guidance not a template. It should read like a short documentation of the feature not like a form !

**Comment by jbescos** [comment]
This is the new issue:
https://github.com/helidon-io/helidon/issues/8772

Are there more changes required in this PR or can it be approved/merged. I guess the documentation changes will be submitted in the new issue I created.

I still don't have any idea where/how should I document this. I will check it with @tvallin because he is in the same time zone than me.

**Review by romain-grecourt — APPROVED** [review]
LGTM, nice work.
I've updated the description.

**Pull request merged**
